### PR TITLE
XNNPACK: Lift constant mul scalars for partitioning

### PR DESCRIPTION
**Pull request by mansnils**
XNNPACK supports the tensor overload for multiply, but aten.mul.Scalar is
 normally decomposed before the XNNPACK partitioner can select it.

Add a MulScalarConfig that preserves eligible scalar multiplications during
partitioning. During XNNPACK preprocessing, lift the scalar into an exported
constant-tensor placeholder and rewrite the operation to aten.mul.Tensor,
allowing the existing multiply serialization path to handle it. This avoids
introducing an aten.full operation and works with the default
to_edge_transform_and_lower flow without explicit XNNPACK transform passes.

Run scalar lifting after SDPA conversion so ConvertToSDPAPass can recover
attention scales from the original aten.mul.Scalar operations. Also allow SDPA
conversion to match graphs processed by the default no-op-expand transform.

For DeIT Tiny, this removes 24 portable aten.mul.Scalar nodes and reduces delegate count from 62 to 50. In current local timing checks the latency impact is modest: about 1% faster on both Android SME2 and the aarch64 XNNPACK/KleidiAI NEON-class host runner. These are modest uplifts but may introduce more opportunities for improvements.

cc @GregoryComer @digantdesai @cbilgin @freddan80 @per @zingo @oscarandersson8218 @Sebastian-Larsson @robell @rascani

### CODE REVIEW

**Comment by mansnils** [comment]
@claude review this PR

File: backends/xnnpack/test/tester/tester.py [hunk 1]
```diff
@@ -77,13 +78,29 @@ def __init__(
         self,
         partitioners: Optional[List[Partitioner]] = None,
         edge_compile_config: Optional[EdgeCompileConfig] = None,
+        transform_passes: Optional[List[PassType]] = None,
     ):
```
[the remark this answers is not in the document]
Comment (mansnils) [reply backends/xnnpack/test/tester/tester.py:80]: Done

File: backends/xnnpack/test/tester/tester.py [hunk 2]
```diff
@@ -132,3 +149,28 @@ def __init__(
...
+            )
+        else:
+            if partitioners is not None:
+                to_edge_and_lower_stage.partitioners = partitioners
+            if edge_compile_config is not None:
+                to_edge_and_lower_stage.edge_compile_conf = edge_compile_config
+        return super().to_edge_transform_and_lower(
+            to_edge_and_lower_stage,
+            generate_etrecord=generate_etrecord,
+        )
```
[the remark this answers is not in the document]
Comment (mansnils) [reply backends/xnnpack/test/tester/tester.py]: Done

File: backends/xnnpack/test/tester/tester.py [hunk 3]
```diff
@@ -24,9 +24,10 @@
     QuantizationConfig,
 )
 from executorch.backends.xnnpack.utils.configs import get_xnnpack_edge_compile_config
-from executorch.exir import EdgeCompileConfig
+from executorch.exir import EdgeCompileConfig, to_edge_transform_and_lower
 from executorch.exir.backend.partitioner import Partitioner
 from torch._export.pass_base import PassType
+from torch.export import ExportedProgram
 from torchao.quantization.pt2e.quantizer import Quantizer
```
[the remark this answers is not in the document]
Comment (mansnils) [reply backends/xnnpack/test/tester/tester.py:30]: Done

File: backends/xnnpack/test/passes/test_lift_constant_scalar_operands_pass.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    class AddScalar(torch.nn.Module):
+        def forward(self, x):
+            return torch.ops.aten.add.Scalar(x, 0.5)
+
+    def _to_edge_graph(self, module):
+        edge = to_edge(
+            torch.export.export(module, (torch.randn(2, 3),), strict=True),
+            compile_config=get_xnnpack_edge_compile_config(skip_dim_order=True),
+        )
+        return edge.transform([LiftConstantScalarOperandsPass()]).exported_program()
```
[the remark this answers is not in the document]
Comment (mansnils) [reply backends/xnnpack/test/passes/test_lift_constant_scalar_operands_pass.py]: Done

**Comment by mansnils** [comment]
@claude review this PR

**Comment by mansnils** [comment]
> **Claude finished @mansnils's task in 9m 47s** —— [View job](https://github.com/pytorch/executorch/actions/runs/28176980031)
> ### PR Review: XNNPACK — Lift constant mul scalars for partitioning
> 
> Reviewed the diff against `origin/main`. Overall this is a clean, well-scoped change with good test coverage and a thoughtful SDPA guard. A few points worth considering before merge.
> #### Correctness / design
> 
>     1. **Constant added as module buffer + `get_attr` rather than a lifted constant placeholder** — `lift_constant_scalar_operands_pass.py:55-76` uses `register_buffer` + `graph.get_attr`. In a _lifted_ edge `ExportedProgram`, the idiomatic way to introduce a constant is `create_constant_placeholder` in `backends/transforms/utils.py:98`, which also registers it in `exp_program.constants` / the graph signature. Because this pass subclasses `ExportPass` (GraphModule-only `call()`), it has no access to the `ExportedProgram`, which is presumably why the buffer route was taken.
>        In practice XNNPACK serialization handles raw `get_attr` nodes (`node_visitor.py:577` → `get_param_tensor` has a get-attr fallback at `utils.py:89-94`), so the delegated path works. The risk is the **non-delegated** path: if a mapped `mul.Scalar` is _not_ partitioned, the orphaned `get_attr` has no graph-signature entry and the ExecuTorch emitter may not handle it. The dtype guard (point 3) makes this unlikely for `mul`, but it's worth confirming the XNNPACK CI job covers a case where the lifted mul falls back to portable. Worth a comment explaining why a buffer + `get_attr` is safe here.
> 
>     2. **`_feeds_sdpa_qk_bmm` passthrough set is narrow and coupled to SDPA internals** (`lift_constant_scalar_operands_pass.py:85-110`). The traversal only walks through `expand_copy`/`view_copy` to reach a `bmm`. If the real SDPA graph has any other op (transpose/permute/to_copy) between the scale `mul.Scalar` and the QK^T `bmm`, the guard misses it, the scalar gets lifted, and `ConvertToSDPAPass.get_scale` (`convert_to_sdpa.py:33-52`) silently fails its `assert scale is not None`. The unit test only exercises the canonical pattern from `partition/graphs/sdpa.py`. Since this generic pass now encodes knowledge of the XNNPACK SDPA pattern, consider deriving the guard from the same `sdpa.get_graphs()` source, or at least add a test with a pattern that has an intervening op. A false _positive_ (leaving a non-SDPA mul→bmm scalar unlifted) is only a missed optimization, so the asymmetry is in the safe direction — but a false _negative_ breaks SDPA.
> 
> 
> #### Tester change
> 
>     3. **`transform_passes` silently dropped when a prebuilt stage is passed** (`test/tester/tester.py`, `Tester.to_edge_transform_and_lower`). In the `else` branch (caller supplies `to_edge_and_lower_stage`), `partitioners` and `edge_compile_config` are propagated onto the stage but `transform_passes` is not. Passing both a stage _and_ `transform_passes=` would silently ignore the passes. Either propagate it (`to_edge_and_lower_stage.transform_passes = transform_passes`) for consistency, or document that they're mutually exclusive. [Fix this →](https://claude.ai/code?q=In%20backends/xnnpack/test/tester/tester.py%20Tester.to_edge_transform_and_lower%2C%20the%20else%20branch%20that%20handles%20a%20caller-supplied%20to_edge_and_lower_stage%20propagates%20partitioners%20and%20edge_compile_config%20but%20not%20transform_passes.%20Propagate%20transform_passes%20onto%20the%20stage%20too%20when%20it%20is%20not%20None.&repo=pytorch/executorch)
> 
> 
> #### Minor
> 
>     4. **`self._modified` can be a local** (`lift_constant_scalar_operands_pass.py:53,113,143,149-150`). It's set/reset entirely within `call()` and never read by helpers, so a local `modified` would be simpler and avoids surprising state on a reused pass instance. Per CLAUDE.md ("choose simpler, more concise"), worth collapsing.
> 
>     5. **Production lowering path doesn't pick this up.** `get_transform_passes()` now includes the pass, but the XNNPACK recipe (`recipes/xnnpack_recipe_provider.py:103-106`, `LoweringRecipe`) sets no `edge_transform_passes`, so the standard recipe-based export does not benefit — only callers that explicitly pass `get_transform_passes()` (tests, `size_analysis_tool`). If the DeIT delegate-count/latency wins are meant to reach normal users, consider wiring it into the `LoweringRecipe`. If opt-in is intentional for now, a one-line note in the PR description would help.
> 
>     6. **0-d constant tensor** — `torch.tensor(value, ...)` (`:66`) produces a rank-0 tensor relying on `mul.Tensor` broadcasting. The end-to-end `test_fp32_mul_scalar` covers this, but it's worth confirming the XNNPACK CI job is green since 0-dim operands have historically been finicky in some backends.
> 
> 
> #### Tests
> 
>     * `test/passes/test_lift_constant_scalar_operands_pass.py` nicely covers lift, unmapped-op passthrough, and the SDPA guard.
> 
>     * `test_fp32_mul_scalar` exercises the full export→serialize→run path. Consider adding a case that asserts a _non-canonical_ SDPA-like graph still recovers scale, to lock in point 2.
> 
> 
> Nothing here is blocking on its own; #1 and #2 are the items I'd most want addressed (or explicitly reasoned about in a comment) before merge. · branch `ops`

Done. 
1.
I am not sure that a missed guard hard-fails SDPA lowering.
  create_sdpa() and its assert scale is not None run only after SubgraphMatcher matches an SDPA pattern. That pattern contains aten.mul.Scalar. If this pass
  mistakenly converts the scale op to aten.mul.Tensor, the pattern no longer matches, so conversion is skipped and get_scale() is never called.
2.
Prefer that in a separate PR.

File: backends/xnnpack/_passes/lift_constant_scalar_operands_pass.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            index += 1
+        return f"{prefix}{index}"
+
+    def _feeds_sdpa_qk_bmm(self, node: torch.fx.Node) -> bool:
+        """
+        Return true for the scale muls consumed by XNNPACK's SDPA pattern.
+
+        ConvertToSDPAPass recovers the user-specified attention scale from the
+        pre-QK^T ``aten.mul.Scalar`` nodes. Keep those scalar muls intact so
+        SDPA conversion can still find the scale before replacing the pattern.
```
Comment (digantdesai) [inline backends/xnnpack/_passes/lift_constant_scalar_operands_pass.py]: sigh.. can we fix that pass?

Comment (mansnils) [reply backends/xnnpack/_passes/lift_constant_scalar_operands_pass.py]: I think you mean keep SDPA logic in SDPA pass, which is a good point.
Done

**Comment by JakeStevens** [comment]
seems not to work with the defaulg get_transform_passes()

Can you add following regression test and fix:

```
from executorch.backends.xnnpack.utils.configs import (
      get_transform_passes,
      get_xnnpack_edge_compile_config,
  )

  def test_converts_sdpa_after_default_transform_passes(self):
      q = torch.randn(2, 4, 8, 16)
      k = torch.randn(2, 4, 8, 16)
      v = torch.randn(2, 4, 8, 16)
      mask = torch.randn(8, 8)

      edge = to_edge(
          torch.export.export(self.SDPA(), (q, k, v, mask), strict=True),
          compile_config=get_xnnpack_edge_compile_config(),
      )

      exported_program = ExportedProgramPassManager(get_transform_passes())(
          edge.exported_program()
      ).exported_program

      exported_program = ExportedProgramPassManager(
          [ConvertToSDPAPass(exported_program)]
      )(exported_program).exported_program

      graph = exported_program.graph_module.graph
      self.assertTrue(
          any(
              node.target
              == exir_ops.edge.aten.scaled_dot_product_attention.default
              for node in graph.nodes
          )
      )
```

**Comment by mansnils** [comment]
> seems not to work with the defaulg get_transform_passes()
> 
> Can you add following regression test and fix:
> 
> ```
> from executorch.backends.xnnpack.utils.configs import (
>       get_transform_passes,
>       get_xnnpack_edge_compile_config,
>   )
> 
>   def test_converts_sdpa_after_default_transform_passes(self):
>       q = torch.randn(2, 4, 8, 16)
>       k = torch.randn(2, 4, 8, 16)
>       v = torch.randn(2, 4, 8, 16)
>       mask = torch.randn(8, 8)
> 
>       edge = to_edge(
>           torch.export.export(self.SDPA(), (q, k, v, mask), strict=True),
>           compile_config=get_xnnpack_edge_compile_config(),
>       )
> 
>       exported_program = ExportedProgramPassManager(get_transform_passes())(
>           edge.exported_program()
>       ).exported_program
> 
>       exported_program = ExportedProgramPassManager(
>           [ConvertToSDPAPass(exported_program)]
>       )(exported_program).exported_program
> 
>       graph = exported_program.graph_module.graph
>       self.assertTrue(
>           any(
>               node.target
>               == exir_ops.edge.aten.scaled_dot_product_attention.default
>               for node in graph.nodes
>           )
>       )
> ```

Should be working now. Done

**Comment by JakeStevens** [comment]
Can you add a root-graph guard in MulScalarConfig.check_constraints:
```
def check_constraints(self, node, ep):
      if node.graph is not ep.graph:
          return False

      if not self.check_common_constraints(node, ep):
          return False
...

and an accompanying regression test exporting a torch.cond whose branches contain mul.Scalar? otherwise the hop delegation fails

**Comment by mansnils** [comment]
> Can you add a root-graph guard in MulScalarConfig.check_constraints:
> 
> ```
> def check_constraints(self, node, ep):
>       if node.graph is not ep.graph:
>           return False
> 
>       if not self.check_common_constraints(node, ep):
>           return False
> ...
> 
> and an accompanying regression test exporting a torch.cond whose branches contain mul.Scalar? otherwise the hop delegation fails
> ```

Yes good catch, done

**Review by JakeStevens — APPROVED** [review]
